File: backend/libs/go/grpc/request/pipeline/steps/validation.go
package steps

import (
	"buf.build/go/protovalidate"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
	"github.com/stigmer/stigmer/backend/libs/go/grpc/request/pipeline"
	"google.golang.org/protobuf/proto"
)

// ValidateProtoStep validates protobuf field constraints using protovalidate.
// This step ensures that the input message satisfies all validation rules
// defined in the proto file (e.g., required fields, min/max values, regex patterns).
//
// Field validation is also enforced at the transport boundary by the
// protovalidate interceptor, so on the gRPC path this step is a redundant
// (harmless) second check. It is retained deliberately because it is the ONLY
// validation on the direct-Go-call path — unit tests that call controllers
// directly rely on it — and both layers share one validator (SharedValidator),
// so the overlap costs nothing.
type ValidateProtoStep[T proto.Message] struct {
	validator protovalidate.Validator
}

// NewValidateProtoStep creates a new validation step backed by the process-wide
// shared validator (constraints are compiled once, not per request).
func NewValidateProtoStep[T proto.Message]() *ValidateProtoStep[T] {
	return &ValidateProtoStep[T]{validator: grpclib.SharedValidator()}
}

// Name returns the step name for logging and tracing.
func (s *ValidateProtoStep[T]) Name() string {
	return "ValidateProtoConstraints"
}

// Execute validates the input message against its proto validation rules.
func (s *ValidateProtoStep[T]) Execute(ctx *pipeline.RequestContext[T]) error {
	if err := s.validator.Validate(ctx.Input()); err != nil {
		return grpclib.InvalidArgumentError("%v", err)
	}
	return nil
}

File: backend/services/stigmer-server-ts/src/pipeline/steps/validation.ts
/**
 * ValidateProto — ports steps/validation.go. Redundant on the RPC path
 * (the protovalidate interceptor already validated at the boundary) but
 * deliberately retained: it is the ONLY validation on the direct-call path
 * unit tests use, and the shared validator makes the overlap free.
 */
import { Code, ConnectError } from "@connectrpc/connect";
import type { DescMessage } from "@bufbuild/protobuf";
import { createValidator } from "@bufbuild/protovalidate";
import type { Validator } from "@bufbuild/protovalidate";

import type { PipelineStep } from "../pipeline.js";
import type { RequestContext } from "../request-context.js";

// Process-wide shared validator (Go SharedValidator): constraints compile
// once, not per request.
let sharedValidator: Validator | undefined;

export function validator(): Validator {
  if (sharedValidator === undefined) {
    sharedValidator = createValidator();
  }
  return sharedValidator;
}

export function newValidateProtoStep<Desc extends DescMessage>(): PipelineStep<Desc> {
  return {
    name: "ValidateProto",
    execute(ctx: RequestContext<Desc>): void {
      const result = validator().validate(ctx.schema, ctx.input);
      if (result.kind === "valid") {
        return;
      }
      if (result.kind === "invalid") {
        throw new ConnectError(result.error.message, Code.InvalidArgument);
      }
      // A rule that fails to compile/evaluate is a server defect, not bad
      // input (same mapping as the transport interceptor).
      throw new ConnectError(
        `validation could not run: ${result.error.message}`,
        Code.Internal,
      );
    },
  };
}
